installinator/src/write.rs

```diff
@@ -0,0 +1,466 @@
+                .send(ReportEvent::Completion(
+                    CompletionEventKind::WriteFailed {
+                        attempt,
+                        kind: artifact_id.kind.clone(),
+                        destination: destination.to_owned(),
+                        written_bytes,
+                        total_bytes,
+                        elapsed: start.elapsed(),
+                        message: format!("flush failed: {error}"),
+                    },
+                ))
+                .await;
+            return Err(error).with_context(|| {
+                format!(
+                    "failed to flush artifact {artifact_id:?} \
+                     ({total_bytes} bytes) to destination `{destination}`"
+                )
+            });
+        }
+    };
+
+    let _ = event_sender
+        .send(ReportEvent::Completion(CompletionEventKind::WriteCompleted {
+            attempt,
+            kind: artifact_id.kind.clone(),
+            destination: destination.to_owned(),
+            artifact_size: total_bytes,
+            elapsed: start.elapsed(),
+        }))
+        .await;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::VecDeque;
+
+    use super::*;
+    use crate::test_helpers::{dummy_artifact_hash_id, with_test_runtime};
+
+    use anyhow::Result;
+    use bytes::{Buf, Bytes};
+    use camino::Utf8Path;
+    use futures::StreamExt;
+    use omicron_test_utils::dev::test_setup_log;
+    use partial_io::{
+        proptest_types::{
+            interrupted_would_block_strategy, partial_op_strategy,
+        },
+        PartialAsyncWrite, PartialOp,
+    };
+    use proptest::prelude::*;
+    use tempfile::tempdir;
+    use test_strategy::proptest;
+    use tokio::io::AsyncReadExt;
+    use tokio_stream::wrappers::ReceiverStream;
+
+    #[proptest(ProptestConfig { cases: 32, ..ProptestConfig::default() })]
+    fn proptest_write_artifact(
+        #[strategy(prop::collection::vec(prop::collection::vec(any::<u8>(), 0..8192), 0..16))]
+        data: Vec<Vec<u8>>,
+        #[strategy(op_strategy())] partial_ops: VecDeque<Vec<PartialOp>>,
+    ) {
+        with_test_runtime(move || async move {
+            proptest_write_artifact_impl(data, partial_ops)
+                .await
+                .expect("test failed");
+        })
+    }
+
+    fn op_strategy() -> impl Strategy<Value = VecDeque<Vec<PartialOp>>> {
+        // XXX: Ideally we'd be able to figure out how many bytes get written by
+        // merely inspecting the list of operations, but partial-io is a bit
+        // broken:
+        // https://github.com/sunshowers-code/partial-io/issues/32
+        //
+        // For now, always error out on earlier attempts and have one successful
+        // attempt at the end. Revisit this after fixing upstream.
+        let success_strategy = prop::collection::vec(
+            partial_op_strategy(interrupted_would_block_strategy(), 1024),
+            0..16,
+        );
```

Comment:
This is pretty annoying, something I want to fix in partial-io upstream after PVT1.